### ISSUE TITLE
[44] Self-Attention does not need O(n^2) Memory

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#72]
[paper](https://arxiv.org/abs/2112.05682)

sequence length 기준으로 
time complexity 는 O(n^2) 그대로 두고,
memory 는 O(n^2) 에서 O(log n) 을 만들어 버린 논문.
실질적으로 구현하면 O(sqrt n) 이 나온다고 함.

large scale 실험들이 난무하는 가운데, 오아시스 역할을 해 줄 논문. 

16384 length 기준 메모리가 inference 에서 59배 감소, training 에서 32배 감소

# Algorithm
원래 single-query attention은 다음과 같음
![image](https://user-images.githubusercontent.com/16400591/146379943-61134aaf-11c6-481e-8ed1-61598d719c82.png)

분배법칙을 쓰면 softmax 에서 쓰던 나눗셈을 다음과 같이 옮겨 버릴 수 있음.
![image](https://user-images.githubusercontent.com/16400591/146380040-d8a87cd3-4361-46a8-b74c-85a56832c775.png)

softmax normalize를 한 다음 attention을 계산하는 것이 아니라 unnormalized score에 대해 계산한 다음 normalize를 하는 식으로 순서를 바꾼 것.

물론 unnormalized score에 대해 그냥 계산하면 값이 터지니 max 값 추적도 같이 들어감.

개인적으로는, 코드를 보니까 더 잘 와닿았음.

```python
import functools
import math

import jax
from jax import numpy as jnp


def _query_chunk_attention(query, key, value, precision, key_chunk_size=4096):
    """Multi-head dot product attention with a limited number of queries."""
    num_kv, num_heads, k_features = key.shape
    v_features = value.shape[-1]
    key_chunk_size = min(key_chunk_size, num_kv)
    query = query / jnp.sqrt(k_features)  # attn 에서 sqrt 로 나눠주는건 국룰이지.

    @functools.partial(jax.checkpoint, prevent_cse=False)
    def summarize_chunk(query, key, value):
        # 이곳이 메인
        # attnetion weight 는 "d" dim 사라지며 einsum.
        # 사실상 dot-product 밖에 안 한 상태
        attn_weights = jnp.einsum("qhd,khd->qhk", query, key, precision=precision)

        # max attn 추적
        max_score = jnp.max(attn_weights, axis=-1, keepdims=True)
        max_score = jax.lax.stop_gradient(max_score)

        # 이제 max attn score 를 빼주면서 값 안정화를 시켜줌
        # exp() 를 통과시켜도 터지지 않는다 !! (attn_weights - max_score < 0)
        exp_weights = jnp.exp(attn_weights - max_score)

        # exp weight 구한 걸로 value 에 먼저 dot product. loop 돌면서 memory 감소.
        exp_values = jnp.einsum("vhf,qhv->qhf", value, exp_weights, precision=precision)
        return (
            exp_values,
            exp_weights.sum(axis=-1),
            max_score.reshape((query.shape[0], num_heads)),
        )

    def chunk_scanner(chunk_idx):
        # key, value 도 4096으로 chunk 떠버림
        key_chunk = jax.lax.dynamic_slice(
            key, (chunk_idx, 0, 0), slice_sizes=(key_chunk_size, num_heads, k_features)
        )
        value_chunk = jax.lax.dynamic_slice(
            value,
            (chunk_idx, 0, 0),
            slice_sizes=(key_chunk_size, num_heads, v_features),
        )
        return summarize_chunk(query, key_chunk, value_chunk)

    # 여기서 사실상, exp-value까지 전부 구했으니, 남은 건 softmax 에서 빼낸 나누기 term
    chunk_values, chunk_weights, chunk_max = jax.lax.map(
        chunk_scanner, xs=jnp.arange(0, num_kv, key_chunk_size)
    )

    # chunk 별 max_attn score 는 이미 추적해 놓았다구. 최대 attn_score 는 몇 인지 구함.
    global_max = jnp.max(chunk_max, axis=0, keepdims=True)

    # chunk-max. shape (num_chunks,) 인 녀석에서 global_max 를 빼봄
    max_diffs = jnp.exp(chunk_max - global_max)

    # difference 만큼 exp_values랑 weights 에 값을 곱해줌.
    chunk_values *= jnp.expand_dims(max_diffs, axis=-1)
    chunk_weights *= max_diffs

    # 값 보정까지 마쳤으니, 최종 attention 계산.
    all_values = chunk_values.sum(axis=0)
    all_weights = jnp.expand_dims(chunk_weights, -1).sum(axis=0)
    return all_values / all_weights


def attention(
    query, key, value, precision=jax.lax.Precision.HIGHEST, query_chunk_size=1024
):
    """Memory-efficient multi-head dot product attention."""
    num_q, num_heads, q_features = query.shape  # (B, nH, dim)

    def chunk_scanner(chunk_idx, _):
        # 1024 단위로 chunk 내버림
        query_chunk = jax.lax.dynamic_slice(
            query,
            (chunk_idx, 0, 0),
            slice_sizes=(min(query_chunk_size, num_q), num_heads, q_features),
        )

        # chunked query 는 chunk_attention 에 들어감
        return (
            chunk_idx + query_chunk_size,
            _query_chunk_attention(query_chunk, key, value, precision=precision),
        )

    _, res = jax.lax.scan(
        chunk_scanner, init=0, xs=None, length=math.ceil(num_q / query_chunk_size)
    )
    return res.reshape(num_q, num_heads, value.shape[-1])

```

# Results
![image](https://user-images.githubusercontent.com/16400591/146387119-96195f38-e24f-4482-a6f7-900e9cb6bcce.png)
![image](https://user-images.githubusercontent.com/16400591/146387172-284149ba-ce10-47d2-a01a-4b2f10d124ca.png)
